File: vn/src/main/java/org/sqlbuilder/vn/objects/Role.java
package org.sqlbuilder.vn.objects;

import org.sqlbuilder.common.*;

import java.util.Comparator;
import java.util.Objects;

public class Role implements Insertable, HasId
{
	public static final Comparator<Role> COMPARATOR = Comparator.comparing(Role::getClazz).thenComparing(Role::getRestrRole);

	public static final SetCollector<Role> COLLECTOR = new SetCollector<>(COMPARATOR);

	private final RestrainedRole restrainedRole;

	private final VnClass clazz;

	// C O N S T R U C T O R

	public static Role make(final VnClass clazz, final RestrainedRole restrainedRole)
	{
		var m = new Role(clazz, restrainedRole);
		COLLECTOR.add(m);
		return m;
	}

	private Role(final VnClass clazz, final RestrainedRole restrainedRole)
	{
		this.restrainedRole = restrainedRole;
		this.clazz = clazz;
	}

	// A C C E S S

	public RestrainedRole getRestrRole()
	{
		return restrainedRole;
	}

	public RoleType getRoleType()
	{
		return restrainedRole.getRoleType();
	}

	public VnClass getClazz()
	{
		return clazz;
	}

	@ProvidesIdTo(type = Role.class)
	@Override
	public Integer getIntId()
	{
		return COLLECTOR.get(this);
	}

	// I D E N T I T Y

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Role that = (Role) o;
		return restrainedRole.equals(that.restrainedRole) && clazz.equals(that.clazz);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(restrainedRole, clazz);
	}

	// I N S E R T

	@RequiresIdFrom(type = VnClass.class)
	@RequiresIdFrom(type = RoleType.class)
	@RequiresIdFrom(type = Restrs.class)
	@Override
	public String dataRow()
	{
		Restrs restrs = restrainedRole.getRestrs();
		Integer restrsid = restrs == null ? null : restrs.getIntId();
		return String.format("%d,%d,%s", clazz.getIntId(), //
				restrainedRole.getRoleType().getIntId(), //
				Utils.nullableInt(restrsid) //
		);
	}

	@Override
	public String comment()
	{
		var r = restrainedRole.getRestrs();
		return String.format("%s,%s,%s", clazz.getName(), restrainedRole.getRoleType().getType(), //
				r == null ? "∅" : r.toString());
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return "class=" + clazz + " role=" + restrainedRole;
	}
}

File: vn/src/main/java/org/sqlbuilder/vn/objects/Role.kt
package org.sqlbuilder.vn.objects

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.HasId
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.common.SetCollector
import org.sqlbuilder.common.Utils.nullableInt
import java.util.*

class Role private constructor(
    val clazz: VnClass,
    val restrRole: RestrainedRole
) : Insertable, HasId {

    @Suppress("unused")
    val roleType: RoleType
        get() = restrRole.roleType

    @RequiresIdFrom(type = Role::class)
    override fun getIntId(): Int {
        return COLLECTOR.invoke(this)
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as Role
        return restrRole == that.restrRole && clazz == that.clazz
    }

    override fun hashCode(): Int {
        return Objects.hash(restrRole, clazz)
    }

    // I N S E R T

    @RequiresIdFrom(type = VnClass::class)
    @RequiresIdFrom(type = RoleType::class)
    @RequiresIdFrom(type = Restrs::class)
    override fun dataRow(): String {
        val restrs = restrRole.restrs
        val restrsid = restrs?.intId
        return "${clazz.intId},${restrRole.roleType.intId},${nullableInt(restrsid)}"
    }

    override fun comment(): String {
        val r = restrRole.restrs
        return  "${clazz.name}, ${restrRole.roleType.type},${r}"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "class=$clazz role=$restrRole"
    }

    companion object {

        val COMPARATOR: Comparator<Role> = Comparator
            .comparing<Role, VnClass> { it.clazz }
            .thenComparing<RestrainedRole> { it.restrRole }

        val COLLECTOR: SetCollector<Role> = SetCollector<Role>(COMPARATOR)

        fun make(clazz: VnClass, restrainedRole: RestrainedRole): Role {
            val m = Role(clazz, restrainedRole)
            COLLECTOR.add(m)
            return m
        }
    }
}
